### ISSUE TITLE
Ubuntu ppa

### ISSUE THREAD

**Issue by DoctorJellyface** [OtterBrowser/otter-browser#352]
Final thing today! (if you have anything against me posting so much issues, just nod. I just want to make Otter better)

Could there be an Ubuntu ppa (Personal Package Archive) for Otter, so that people could install and update through apt-get?

In the best case, 2 ppa's:

otter-stable (for non-weekly builds)
otter-daily (for the weekly builds, following the launchpad terminology)

Thanks.

**Comment by Emdek** [comment]
@landroni, I guess that we should finally revisit this ticket. ;-)

**Comment by landroni** [comment]
Sure. I've been using the daily builds all this time and it works nicely in my experience. If you want to switch to weekly builds, then change schedule to "manual" triggering of the builds (you hit the button only when a weekly release has been committed). If you need to have changes to the code that don't actually make the repo, then you may consider creating a "release" branch that gets updated only when necessary and diverges from master quite a bit, once every release. 

You can of course also switch to manual packaging, and I can fill you in, but such packaging as you know is much more involved... Just let me know.

**Comment by landroni** [comment]
I've now updated the stable PPA to 0.9.05 beta.

**Comment by Emdek** [comment]
@landroni, as discussed earlier, I would like to keep daily PPA. ;-)
Weeklies require changes that aren't sent to master, changes in project files to show correct version, so separate branch will be required.

BTW, are these beta packages built using tag or hotfix packages?

**Comment by landroni** [comment]
> BTW, are these beta packages built using tag or hotfix packages?

@Emdek Come again? :)

**Comment by Emdek** [comment]
@landroni, there were new source packages released next day, fixing one important crash, available on sf.net. ;-)

**Comment by DoctorJellyface** [comment]
Why don't you just create a stable, snapshot (weekly) and daily PPA?

**Comment by landroni** [comment]
@Emdek Oh, that thing... :) Thanks for the heads-up. 

BTW, why not simply do a minor bump to the release: 0.9.05.1? This is kinder on package maintainers, it's clearer for users, and allows you to have a 2nd consecutive hotfix if need be: 0.9.05.2 

In any case, this is what other projects tend to do when there is a new minor release fixing a single issue.

**Comment by landroni** [comment]
@DoctorJellyface 

We have the stable (if beta releases pass for stable these days) and daily PPAs. Both are low-maintenance. 

A weekly PPA would require significantly more attention to maintain. It would require either doing manual releases each week (which I'm not quite ready to take up), or setting up a separate "weekly" branch which gets updated upon releases (which @Emdek  doesn't seem too enthusiastic on). The daily works quite well for me, so I'll leave it at that for the time being. 

If anyone is interested in "weekly" builds your best bet is to install the daily build that was published the day after the weekly release...

**Comment by Emdek** [comment]
@landroni, yeah, that was first minute fixing for last minute regression, due to that "extreme week" with more than 80 commits. ;-)
Hopefully there will be simply more time to spot issues before tagging, making use of @OtterBrowser/qa. ;-)

I can try to work on weeklies, is there any short guide? ;-)
Previously I was maintaining builds with use of CMake and VM for 32 bit variant,

**Comment by landroni** [comment]
@Emdek I've now uploaded the hotfix packackages using the 0.90.5.1 tag. 

Re weeklies... The simplest solution would be to switch the daily builds to a "manual" schedule, create a new branch that is synced with master *only when* a new weekly release commit was performed in master (very easy and no extra-maintenance). We would avoid updating the meta info, thus keeping the daily builds tagging scheme, but all the relevant builds would then literally be synced with the weekly releases (apart from the extraneous tags and metadata). 

Again, IMO this would be the easiest approach. If you do prefer properly released weeklies, I will shoot out the instructions then...

**Comment by Emdek** [comment]
@landron, I would like to compare required effort by checking instructions, if it is not a big problem. ;-)

**Comment by landroni** [comment]
OK, here we go. 

**Step 0.**
- Before anything else you need to have set up Launchpad, your keys SSH and OpenPGP, so that you can upload stuff. Let me know if you have trouble figuring it out.

**Step 1.**
- You usually take the `*.debian.tar.gz` of the previous version (for your Ubuntu distro) and the  release archive of the new version. See also:
 http://askubuntu.com/questions/603157/what-is-a-orig-tar-xz-file-in-packaging

- Rename source archive to: `<name>_<version>.orig.tar.gz`, then unpack in a dir of your choice (I usually do this in a dir called 'ppa0').

- Unpack `*.debian.tar.gz` into the unpacked `*.orig.tar.gz`.

**Step 2.** 
Once this is done, you need to bump the version. What I now do is:

    dch -v "0.9.06-1~trusty~ppa1" "New upstream release" && dch -r --distribution trusty ""

See also: 
http://askubuntu.com/questions/592322/how-to-easily-bump-a-package-for-a-ppa

You basically need to create a new entry in the `debian/changelog` (the above command does it for you), adjusting version number and adjusting the date. You need to use your personal credentials (name & email), corresponding to those registered on Launchpad. What goes after the dash `1~trusty~ppa1` is an "internal" versioning scheme for your uploads. Make a typo, something goes wrong with the build, and you bump that to `1~trusty~ppa2`, etc.

**Step 3.** 

Once this is done, you usually want to check that all builds as  expected. For this I clone `ppa0` (think pristine source) to `ppa1` (think messy playground). 

- Then in `ppa1`, to build binaries, from within root of unpacked dir:

     `debuild -b -us -uc`

- In `ppa0`, to build source archive (and include full source code, for your first upload):

     `debuild -S -sa`

- In `ppa0`, to build source archive (without full source code, for remaining distros):

     `debuild -S -sd`

**Step 4.**

Check `*.build` file for errors/issues. Important step until you get used to how it works. 

Then you upload the resulting `*.changes` file to Launchpad using:

    dput ppa:otter-browser/release *.changes


===

It's intimidating at first, and a lot of steps, but it's easier than it sounds once all is set up. What's unnerving is that you need to repeat some of these steps (although there are shortcuts) when dealing with 4 Ubuntu distros at a time. Usually you do your first upload with the full source code (usually for the distro you're using yourself, assuming you tested compilation locally), and remaining uploads without full source code. 

Feel free to shoot any questions that you might have. 

PS Should we put this in the wiki?

**Comment by Emdek** [comment]
@landroni, thanks for recipe, I'll try it tomorrow morning for beta, I wanted to try it earlier but was too busy. :-/
I guess that it makes sense to put it on the wiki, so it won't get lost when this ticket will be finally marked as fixed.
I could be either new page or part of this one:
https://github.com/OtterBrowser/otter-browser/wiki/Build-From-Source

**Comment by Emdek** [comment]
And it fails on importing GPG key (doesn't like fingerprint)... :-/

**Comment by landroni** [comment]
@Emdek 

Can't be of too much help there, other than suggesting to try again and read all instructions on Launchpad you can find. First time I had to do this I fingerprinting business I succeeded by trial and error. 

One thing that may be important is to define an SSH key. I think at first something wasn't working because I hadn't done that. I think you will also need to sign Ubuntu's Code of Conduct.

**Comment by Emdek** [comment]
@landroni, the issue was that my OpenPGP key wasn't exported to external server, I've figured that out with help from one of users of our PPA. ;-)
Now I'm on step 3, ``debuild``, thanks again for recipe.

**Comment by Emdek** [comment]
@landroni, uhm, I think that we overlapped a bit, I haven't noticed that you uploaded new package already, sorry.

BTW, this link may be useful too:
http://blog.gauner.org/blog/2010/12/06/debsign-clearsing-failed-secret-key-not-available/

**Comment by landroni** [comment]
> uhm, I think that we overlapped a bit, I haven't noticed that you uploaded new package already

Upon popular demand... :) 

But that's not an issue. You can play around on a personal, testing PPA until you get the whole process working...

**Comment by Emdek** [comment]
@landroni, I think that I've messed it up a bit, ``dput`` fails to upload source code after executing ``debuild -S -sa`` and then ``debuild -S -sd``, any suggestions?

**Comment by landroni** [comment]
Try this: 

     debuild -S -sa


For the first distro (e.g. utopic). Then `dput` the `*.changes` file. 

Then: 

    debuild -S -sd

*after* having changed the distro in debian/changelog (e.g. to trusty). 

You can redo these steps if you've already done them once. If it still doesn't, please post what you get in the console.

**Comment by Emdek** [comment]
@landroni, ah, I've should read last part more carefully, as it mentions that changelog entry needs to be modified and each distribution needs separate ``dput``, sorry for this commotion.

**Comment by Emdek** [comment]
And now *weekly* PPA is up and running. :-)

**Comment by landroni** [comment]
@Emdek 

I noticed that there weren't RC releases for Trusty (pre-historic, but still supported by Ubuntu). Any plans on making latest release available for Trusty as well?

**Comment by Emdek** [comment]
@landroni, nah, my VM died while pushing packages, I'm preparing new one.
